Enabling the cancellation of non-actor tasks in a worker's queue

gabrieleoliaro opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

These changes are needed to fix issue [#12080](https://github.com/ray-project/ray/issues/12080). 

With the new code, `CoreWorker::HandleCancelTask` is able to cancel non-actor tasks that are waiting (in a worker's queue) for their turn to execute. The changes do not affect how actor tasks are cancelled.


<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

ijrsvt commented on src/ray/core_worker/core_worker.cc:
```diff
@@ -2167,19 +2167,26 @@ void CoreWorker::HandleCancelTask(const rpc::CancelTaskRequest &request,
+  } else if (!kill_success) {
+    
+    // If the task is not currently running, check if it is in the worker's queue of normal tasks, and remove it if found. 
+    if (direct_task_receiver_->CancelQueuedNormalTask(task_id)) {
+      success=true;
+    }
```
Can you just assign?  `success = direct_task_receiver_->CancelQueuedNormalTask(task_id)`?

ijrsvt commented on src/ray/core_worker/core_worker.cc:
```diff
@@ -2167,19 +2167,26 @@ void CoreWorker::HandleCancelTask(const rpc::CancelTaskRequest &request,
   absl::MutexLock lock(&mutex_);
   TaskID task_id = TaskID::FromBinary(request.intended_task_id());
-  bool success = main_thread_task_id_ == task_id;
+  bool kill_success = main_thread_task_id_ == task_id;
```
Could you rename this to `requested_task_running` (i.e. `requested_task_running= main_thread_task_id_ == task_id;`

ijrsvt commented on src/ray/core_worker/core_worker.cc:
```diff
@@ -2167,19 +2167,26 @@ void CoreWorker::HandleCancelTask(const rpc::CancelTaskRequest &request,
+  if (kill_success && !request.force_kill()) {
     RAY_LOG(INFO) << "Interrupting a running task " << main_thread_task_id_;
-    success = options_.kill_main();
+    kill_success = options_.kill_main();
```
This should just be `success`

ijrsvt commented on src/ray/core_worker/transport/direct_actor_transport.h:
```diff
@@ -383,6 +389,11 @@ class ActorSchedulingQueue : public SchedulingQueue {
     ScheduleRequests();
   }
 
+  // We don't cancel actor tasks by removing them from the actor queues, so we don't do anything here.
```
nit: We don't cancel any actor tasks, so maybe:
```
// We don't cancel actor tasks, so we don't do anything here. 
```

ijrsvt reviewed (changes requested):
Looks great! Thanks so much for doing this really quickly. I left a few comments and was also thinking it would be good to add a few CPP tests for this!

stephanie-wang commented on src/ray/core_worker/transport/direct_actor_transport.h:
```diff
@@ -383,6 +389,9 @@ class ActorSchedulingQueue : public SchedulingQueue {
   }
 
+  // We don't cancel actor tasks, so we don't do anything here.
+  bool CancelTaskIfFound(TaskID task_id) { return false; }
```
```suggestion
  bool CancelTaskIfFound(TaskID task_id) { RAY_CHECK(false) << "Cannot cancel actor tasks"; }
```

stephanie-wang reviewed (approved):
Thanks, this looks great! Just left a suggestion to make cancelling an actor task fatal.

stephanie-wang commented:
+1 on adding a unit test!

gabrieleoliaro commented:
@ijrsvt @stephanie-wang I just added two unit tests!

ijrsvt reviewed (approved):
Looks good! Thanks for adding some unit tests :)

stephanie-wang commented:
Looks like there is a related CI test timeout:
```
==================== Test output for //python/ray/tests:test_cancel:
============================= test session starts ==============================
platform linux -- Python 3.6.12, pytest-5.4.3, py-1.9.0, pluggy-0.13.1 -- /opt/miniconda/bin/python3
cachedir: .pytest_cache
rootdir: /home/travis/.cache/bazel/_bazel_travis/b88c129a127452fc94033a29d9f90e20/execroot/com_github_ray_project_ray/bazel-out/k8-opt/bin/python/ray/tests/test_cancel.runfiles/com_github_ray_project_ray
plugins: asyncio-0.14.0, rerunfailures-9.1.1, sugar-0.9.4, timeout-1.4.2
collecting ... collected 15 items

::test_cancel_chain[True] PASSED                                         [  6%]
::test_cancel_chain[False] PASSED                                        [ 13%]
::test_cancel_multiple_dependents[True] PASSED                           [ 20%]
::test_cancel_multiple_dependents[False] PASSED                          [ 26%]
::test_single_cpu_cancel[True] PASSED                                    [ 33%]
::test_single_cpu_cancel[False] PASSED                                   [ 40%]
::test_comprehensive[True] PASSED                                        [ 46%]
::test_comprehensive[False] PASSED                                       [ 53%]
::test_stress[True] -- Test timed out at 2020-11-28 07:07:09 UTC --
```

Also, could you fix the lint (`./ci/travis/format.sh`)?

gabrieleoliaro commented:
I'm working on the timeouts! When it comes to linting, I'm confused as to why the test fails, because the linter script seems to show that the code is already lint-ed:

`(myenv) dhcp-10-250-248-49:ray Gabriele$ ./ci/travis/format.sh 
WARNING: Ray uses mypy 0.782, You currently are using 0.790. This might generate different results.
WARNING: clang-format is not installed!
WARNING: Ray uses flake8 with flake8_quotes. Might error without it. Install with: pip install flake8-quotes
From github.com:ray-project/ray
 * branch                master     -> FETCH_HEAD
(myenv) dhcp-10-250-248-49:ray Gabriele$ git status
On branch cancel_queued_tasks
Your branch is up to date with 'origin/cancel_queued_tasks'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	python/ray/dashboard/

nothing added to commit but untracked files present (use "git add" to track)
`

gabrieleoliaro commented:
@ijrsvt @stephanie-wang Does anyone have any advice on how to debug the `test_cancel` timeout? In the Travis build which fails, it looks like the timeout is caused by `test_stress[True]` within `/python/ray/tests/test_cancel.py`. The` test_cancel.py` pytest shows some weird behavior on my local machine. In particular, `test_stress[True]` always passes (both on the PR branch as well as on the master branch), whereas `test_fast[False]` always gets stuck (it never runs to completion, with low cpu/memory utilization). I'm particularly confused by the fact that `test_fast[False]` doesn't finish even when it is run on master:
<img width="1076" alt="test_cancel" src="https://user-images.githubusercontent.com/6480808/101117970-2e2db700-35b6-11eb-980b-81f7e27a8be3.png">

I can try merging the master branch in again, with `git merge master`, but it looks like the latest commits on ray-project:master also have some Travis builds that fail, so idk if it would help.

The good news is that at least the `CI / windows-msvc (pull_request)` tests are passing now (see the "linting" commit)!

Thanks!

stephanie-wang commented:
It's possible `test_fast` was already flaky. I did see one test failure on flakey-tests.ray.io but it seems to fail infrequently. I think `test_stress` is an actual regression since it's timing out on all of the builds for this run. Sometimes it's difficult to reproduce Travis logs, but you could try running the pytest command with `taskset` to pin to one core. Something like this:
`taskset 1 pytest python/ray/tests/test_cancel.py::test_stress[True]`

@ijrsvt, any pointers on these test cases?

gabrieleoliaro commented:
Yeah, looking at the code again, test_stress seems likely to be caused by a deadlock. On the other hand, I'm still super confused about test_fast. Let me try running the test again with the `taskset` option

ijrsvt commented:
Hmm, that is weird! I am able to reproduce the problem with `test_fast[False]` locally. It appears to hang in the final getting of results.

From some poking around it appears that the failure is in the first `100` results, which means that somehow a task is getting stuck?
I also noticed that while getting here (adding print statements to `test_fast`) that certain print statements made the test more or less flakey.

stephanie-wang commented:
@gabrieleoliaro @ijrsvt looks like `test_cancel` is still timing out. Any ideas why?

gabrieleoliaro commented:
> @gabrieleoliaro @ijrsvt looks like `test_cancel` is still timing out. Any ideas why?

It looks like there is some race condition involving the mutex in the `NormalSchedulingQueue` class. In fact, when I commented it out, all tests in `test_cancel` passed quickly. Let me see if the race condition involves all the methods in NormalSchedulingQueue or only some

stephanie-wang commented:
> It looks like there is some race condition involving the mutex in the `NormalSchedulingQueue` class. In fact, when I commented it out, all tests in `test_cancel` passed quickly. Let me see if the race condition involves all the methods in NormalSchedulingQueue or only some

Ah I see, one tip is if you can reproduce the race condition locally, you can use `gdb` to attach to a hanging process and use `backtrace` to see where it's hanging.

gabrieleoliaro commented:
> > It looks like there is some race condition involving the mutex in the `NormalSchedulingQueue` class. In fact, when I commented it out, all tests in `test_cancel` passed quickly. Let me see if the race condition involves all the methods in NormalSchedulingQueue or only some
> 
> Ah I see, one tip is if you can reproduce the race condition locally, you can use `gdb` to attach to a hanging process and use `backtrace` to see where it's hanging.

let me try that!

gabrieleoliaro commented:
@stephanie-wang @ijrsvt I think I found the problem. While holding the lock, the function NormalSchedulingQueue::ScheduleRequests in direct_actor_transport.h loops through a queue of InboundRequests and calls Accept() on each of them, before popping them from the queue. However, sometimes execution does not come back from this line here: https://github.com/gabrieleoliaro/ray/blob/ff05231dbb3599cb88197ad2c559845f68553391/src/ray/core_worker/transport/direct_actor_transport.h#L569 . So I think I just need to cal head.Accept() without holding the lock.

stephanie-wang commented:
Ah interesting, I'm not sure why calling Accept() would deadlock though. I guess it must be some interaction with the cancellation code. Were you able to figure out where exactly Accept() was hanging? Calling Accept() outside of the lock sounds good, though.

gabrieleoliaro commented:
> Ah interesting, I'm not sure why calling Accept() would deadlock though. I guess it must be some interaction with the cancellation code. Were you able to figure out where exactly Accept() was hanging? Calling Accept() outside of the lock sounds good, though.

I just tested the code after changing the loop in NormalSchedulingQueue::ScheduleRequests() to call Accept() after releasing the lock. While the logs now suggest that the execution no longer gets stuck within Accept(), that threads no longer hang while grabbing a lock or hang/exit/freeze while holding a lock, the tests still get stuck. So I don't think the change fixed the problem (although it might have removed some additional bugs). Will keep investigating! Let me know if you have any guess of what could be the cause. I still think it's some synchronization problem, because everything worked smoothly when I commented out the mutex & locking instructions

gabrieleoliaro commented:
The interesting thing is -- the deadlock seems to be present in the master branch as much as it is in this PR's branch, so it might actually be an issue that is still lingering from the [Queueing non-actor tasks at the workers](https://github.com/ray-project/ray/pull/11051) PR

gabrieleoliaro commented:
@ijrsvt @stephanie-wang Hi everyone, and happy New Year! After experimenting a bit more over the last few days, I am now almost certain that the timeout in the `python/ray/tests/test_cancel.py::test_stress[False]` is independent from the PR. In fact, after noticing that the `test_cancel `pytest was getting stuck locally even with the master branch, I created a new branch (see [here](https://github.com/gabrieleoliaro/ray/tree/revert_PR_11051)) where I reverted the changes from this PR as well as my previous PR #11051). The test still hangs at `python/ray/tests/test_cancel.py::test_stress[False]` (see screenshot below), so the cause cannot be in the new code. 

<img width="927" alt="test_cancel_hangs" src="https://user-images.githubusercontent.com/6480808/103442275-02552e00-4c55-11eb-888d-66e937b2b4f3.png">

After running the `python/ray/tests/test_cancel.py::test_stress[False]` on the new branch (where I had reverted the changes of the two PRs), I also took a look at the core-driver logs and I noticed that they also contained a lot of the errors:

`W1209 00:03:50.420928 69159 422989248 reference_count.cc:195] Tried to decrease ref count for nonexistent object ID: 711f7fe0b401aa0cffffffff0100000001000000`

Hence, these warnings, too, should be independent from the new changes

stephanie-wang commented:
Okay sounds good! I'll merge it for now and keep an eye on the flakey tests. We'll revert this PR if this does increase the flakiness.
